### ISSUE TITLE
"Conv2DBackpropFilter uses a while_loop. Fix that!"

### ISSUE THREAD

**Issue by hnyu** [tensorflow/tensorflow#34496]
**Describe the current behavior**
Currently when I compute Jacobian with the option "experimental_use_pfor=True" on Conv2D layers, the GPU memory will explode. For other layers like FC (with the same amount of parameters), there is no such issue.

**Describe the expected behavior**
If there is any efficient alternative way of doing this (hopefully YES, as indicated by the warning message in the code "Conv2DBackpropFilter uses a while_loop. Fix that!"), please kindly update.

**Code to reproduce the issue**
Call tape.jacobian on a Conv2D network of a moderate size with "experimental_use_pfor=True".

**Comment by gadagashwini** [comment]
@hnyu, Thanks for reporting this issue.
Please provide the complete standalone code to reproduce the reported issue. And also provide the Tensorflow version. Thanks!

**Comment by gadagashwini** [comment]
@hnyu, Could you post your code snippet to analyze the issue.

**Comment by hnyu** [comment]
Sorry for the delay because of the vacation. Below is a minimal reproducible example:

```import tensorflow as tf

if __name__ == "__main__":
    gpus = tf.config.experimental.list_physical_devices('GPU')
    for gpu in gpus:
        tf.config.experimental.set_memory_growth(gpu, True)

    img = tf.zeros([64, 10, 10, 1])
    conv = tf.keras.layers.Conv2D(
        padding="valid",
        filters=5,
        kernel_size=4,
        strides=2)
    deconv = tf.keras.layers.Conv2DTranspose(
        padding="valid",
        filters=1,
        kernel_size=4,
        strides=2)

    with tf.GradientTape() as tape:
        pred_img = deconv(conv(img))
        cost = tf.reduce_sum(tf.square(pred_img - img))

    jacob = tape.jacobian(
        cost,
        conv.trainable_variables + deconv.trainable_variables,
        experimental_use_pfor=True,
        unconnected_gradients=tf.UnconnectedGradients.ZERO)
```

It seems that computing Jacobian using pfor on Conv2DTranspose will throw this message "Conv2DBackpropFilter uses a while_loop. Fix that!", while Conv2D alone will not. I also had a memory explosion problem accompanied with this message (which shouldn't happen because I can compute Jacobian on an MLP with a roughly equal amount of parameters). I suspect the memory issue is related to this warning message, but am not sure. 

But as the warning message suggests, it might be better if it can be addressed.

**Comment by gadagashwini** [comment]
I could replicate the issue with Tenosrflow 2.0 and 2.1-rc0.
Please see the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/8b0694efd70cde49845e2c8745e9610e/untitled296.ipynb). Thanks!

**Comment by agarwal-ashish** [comment]
It's not clear why the while_loop based fallback for handling pfor should be causing the memory increase. The warning here is because the fallback could cause slowdowns compared to a more optimized vectorization. It will be worth looking at some memory profiles to pin down the cause.
Jacobian computation is going to be expensive and memory intensive. It's worth doing some calculations on expected memory growth to see if the observed memory is inline or not.

**Comment by agarwal-ashish** [comment]
It looks like cost is a scalar. Why not use tape.gradient here ? tape.jacobian is expected to be more memory inefficient.
